src/controllers/krate/publish.rs

```diff
@@ -385,16 +385,40 @@ pub async fn publish(app: AppState, req: BytesRequest) -> AppResult<Json<GoodCra
             // Link this new version to all dependencies
             add_dependencies(conn, &deps, version.id)?;
 
-            // Insert the default version if it doesn't already exist. Compared
-            // to only using a background job, this prevents us from getting
-            // into a situation where a crate exists in the `crates` table but
-            // doesn't have a default version in the `default_versions` table.
-            let inserted_default_versions = diesel::insert_into(default_versions::table)
+            let existing_default_version = default_versions::table
+                .inner_join(versions::table)
+                .filter(default_versions::crate_id.eq(krate.id))
+                .select(DefaultVersion::as_select())
+                .first(conn)
+                .optional()?;
+
+            let published_default_version = DefaultVersion {
+                id: version.id,
+                num: semver,
+                yanked: false,
+            };
+
+            // Upsert the `default_value` determined by the existing `default_value` and the
+            // published version. Note that this could potentially write an outdated version
+            // (although this should not happen regularly), as we might be comparing to an
+            // outdated value.
+            //
+            // Compared to only using a background job, this prevents us from getting into a
+            // situation where a crate exists in the `crates` table but doesn't have a default
+            // version in the `default_versions` table.
+            let default_version = if let Some(existing_default_version) = existing_default_version {
+                std::cmp::max(existing_default_version, published_default_version)
+            } else {
+                published_default_version
+            };
+            diesel::insert_into(default_versions::table)
                 .values((
                     default_versions::crate_id.eq(krate.id),
-                    default_versions::version_id.eq(version.id),
+                    default_versions::version_id.eq(default_version.id),
                 ))
-                .on_conflict_do_nothing()
+                .on_conflict(default_versions::crate_id)
+                .do_update()
+                .set(default_versions::version_id.eq(default_version.id))
                 .execute(conn)?;
```

Comment:
```rust
// Upsert the `default_version` determined by the existing `default_version` and the
// published version. Note that this could potentially write an outdated version
// (although this should not happen regularly), as we might be comparing to an
// outdated value.
//
// Compared to only using a background job, this prevents us from getting into a
// situation where a crate exists in the `crates` table but doesn't have a default
// version in the `default_versions` table.

if let Some(existing_default_version) = existing_default_version {
    let published_default_version = DefaultVersion {
        id: version.id,
        num: semver,
        yanked: false,
    };

    if existing_default_version < published_default_version {
        diesel::update(default_versions::table)
            .filter(default_versions::crate_id.eq(krate.id))
            .set(default_versions::version_id.eq(version.id))
            .execute(conn)?;
    }
} else {
    diesel::insert_into(default_versions::table)
        .values((
            default_versions::crate_id.eq(krate.id),
            default_versions::version_id.eq(version.id),
        ))
        .execute(conn)?;
}
```

I guess we could do something like this to skip the update query if the value does not need updating. what do you think?

Reply:
Updated, thanks :)